daemon/api.go

```diff
@@ -2352,6 +2360,80 @@ func postCreateUser(c *Command, r *http.Request, user *auth.UserState) Response
 	}, nil)
 }
 
+type postUserRemoveData struct {
+	Username string `json:"username"`
+}
+
+func postRemoveUser(c *Command, r *http.Request, user *auth.UserState) Response {
+	_, uid, _, err := postRemoveUserUcrednetGet(r.RemoteAddr)
+	if err != nil {
+		return BadRequest("cannot get ucrednet uid: %v", err)
+	}
+	if uid != 0 {
+		return BadRequest("cannot use remove-user as non-root")
+	}
+
+	var removeData postUserRemoveData
+
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&removeData); err != nil {
+		return BadRequest("cannot decode remove-user data from request body: %v", err)
+	}
+
+	// verify request
+	st := c.d.overlord.State()
+	st.Lock()
+	users, err := auth.Users(st)
+	st.Unlock()
+	if err != nil {
+		return InternalError("cannot get user count: %s", err)
+	}
+
+	username := removeData.Username
+	if username == "" {
+		return BadRequest("cannot remove user: 'username' field is empty")
+	}
+
+	// check if user can be removed
+	if _, err := userLookup(username); err != nil {
+		return BadRequest("cannot remove user %s: %s", username, err)
+	}
+	id, err := getUserID(username, users)
+	if err != nil {
+		return BadRequest("cannot remove user %s: user is a system user", username)
+	}
+
+	// TODO: check options
+	opts := &osutil.UserDelOptions{
+		Force:     false,
+		Remove:    true,
+		ExtraUser: !release.OnClassic,
+	}
+
+	// Remove user from system
+	if err := osutilUserDel(username, opts); err != nil {
+		return BadRequest("cannot remove user %s: %s", username, err)
+	}
+	// Remove user from state
+	st.Lock()
```

Comment:
Should we keep the state locked for the duration of this function? I don't mind the extra lock unlock, just wondering.

Reply:
We have a test/execution race there, and also removal from system and from state should be a transaction. In the ideal case I think we should remove from state first and rollback if userdel fails, but I'm not sure if userdel itself is safe -- it could remove user data and the fail when removing the passwd entry or something like that, or even block forever inside the utility.

I didn't keep it locked because in this latter case, access to state would be kept locked forever for everyone. If we lock twice, we have a race situation but only this handler would be blocked (do we have a watchdog for situations like this?). The whole sequence of actions here is not bullet-proof, and it doesn't seem easy to make it so, especially when we don't control the external utility.